elastic/apm-server
# txmetrics: Add global labels to metric sets

## Motivation/summary

Adds global labels to the aggregated metric (as parsed in the `metadata`
event) dimensions that are published by the APM Server.

## Checklist

~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~

## How to test these changes

1. docker-compose up -d
2. cd systemtest && go run cmd/runapm/main.go -f
3. Run the [following program](https://gist.github.com/marclop/a976a2b78b1a90dd8d87b2c31faf67fe)
4. Ensure that the indexed metric documents contain the global labels
```
GET /metrics-apm.internal-default/_search
{
  "_source": false, 
  "fields": ["labels.*"], 
  "query": {"match_all": {}}
}
```
```
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": ".ds-metrics-apm.internal-default-2022.08.10-000001",
        "_id": "4eDLhYIBxUgzjit0xKMG",
        "_score": 1,
        "_source": {},
        "fields": {
          "labels.department_name": [
            "apm"
          ],
          "labels.company": [
            "elastic"
          ],
          "labels.organization": [
            "observability"
          ]
        }
      },
      {
        "_index": ".ds-metrics-apm.internal-default-2022.08.10-000001",
        "_id": "4uDLhYIBxUgzjit0xKMG",
        "_score": 1,
        "_source": {},
        "fields": {
          "labels.department_name": [
            "fleet-server"
          ],
          "labels.company": [
            "elastic"
          ],
          "labels.organization": [
            "observability"
          ]
        }
      },
      {
        "_index": ".ds-metrics-apm.internal-default-2022.08.10-000001",
        "_id": "4-DLhYIBxUgzjit0xKMG",
        "_score": 1,
        "_source": {},
        "fields": {
          "labels.department_name": [
            "kibana"
          ],
          "labels.company": [
            "elastic"
          ],
          "labels.organization": [
            "platform"
          ]
        }
      }
    ]
  }
}
```
5. Enable Service groups in kibana ([link](https://www.elastic.co/guide/en/kibana/8.3/services.html#service-groups))
6. Create a 3 service groups:
   1. Observability: Matches `labels.organization: observability`
   2. APM: Matches `labels.department_name: apm`
   3. Platform: Matches `labels.organization: platform`

7. Verify that Observability has two services (`apm`, `fleet-server`)
8. Verify that APM has 1 service (`apm`)
9. Verify that Platform has 1 service (`kibana`)

## Related issues

Closes #8139
